### ISSUE TITLE
Step functions with API Gateway returning  Could not resolve host

### ISSUE THREAD

**Issue by Vmacedo-Daitan** [localstack/localstack#3777]
# Type of request: This is a ...

[X] bug report
[ ] feature request

# Detailed description
When deploying a step function with and API Gateway the URL created cannot be resolved.
...

## Expected behavior
Call the step functions by API Gateway
...

## Actual behavior
When try to CURL the step functions it returns the error "Could not resolve host"
...

# Steps to reproduce
Create a step function with an HTTP:
```
functions:
  hello:
    handler: handler.hello

stepFunctions:
  stateMachines:
    helloMachine:
      name: hello-machine-step
      events:
        - http:
            path: hello
            method: GET
      definition:
        StartAt: HelloWorld1
        States:
          HelloWorld1:
            Type: Task
            Resource:
              Fn::GetAtt: [hello, Arn]
            End: true
```

After deploying, I get the message below:
```
Service Information
service: stepf-by-api
stage: local
region: us-east-1
stack: stepf-by-api-dev
resources: 12
api keys:
  None
endpoints:
  http://localhost:4566/restapis/2fhrtughr0/local/_user_request_
functions:
  hello: stepf-by-api-local-hello
layers:
  None
Serverless StepFunctions OutPuts
endpoints:
  GET - https://2fhrtughr0.execute-api.us-east-1.localhost/local/hello
```

And when I curl, I get;
'''
> curl https://2fhrtughr0.execute-api.us-east-1.localhost/local/hello
curl: (6) Could not resolve host: 2fhrtughr0.execute-api.us-east-1.localhost
'''

## Command used to start LocalStack
docker-compose up
...

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)

...

**Comment by Vmacedo-Daitan** [comment]
Just a note, when I deployed to AWS it worked fine.

**Comment by amitlevy21** [comment]
Maybe this would help https://github.com/localstack/localstack#invoking-api-gateway

**Comment by Vmacedo-Daitan** [comment]
If I invoke a lambda it works, but for the step functions it doesn't work.
If you look at the deploy results two endpoints were created: one for the lambda and one for the step functions. 
And for the lambda I get the results below:

```
curl -v http://localhost:4566/restapis/7pk96ejbe6/local/_user_request_/hello
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4566 (#0)
> GET /restapis/7pk96ejbe6/local/_user_request_/hello HTTP/1.1
> Host: localhost:4566
> User-Agent: curl/7.55.1
> Accept: */*
>
< HTTP/1.1 500 
< content-type: text/html; charset=utf-8
< content-length: 2
< date: Thu, 01 Apr 2021 13:15:53 GMT
< server: hypercorn-h11
<
{}* Connection #0 to host localhost left intact
```

**Comment by pinzon** [comment]
@Vmacedo-Daitan For the moment, creating the step-function with an http even returns a second endpoint which is not correct, to access the apigateway resources you need to use the first endpoint: ` http://localhost:4566/restapis/7pk96ejbe6/local/_user_request_`

To verify that your step-function resource exists you could use the command  `awslocal apigateway get-resources --rest-api-id $API_ID` and see that it's there

In this case when you call `http://localhost:4566/restapis/7pk96ejbe6/local/_user_request_/hello` you're calling your step function. 

I'll see if we can fix that second endpoint so there's no more confusion.

**Comment by Vmacedo-Daitan** [comment]
Hi! @pinzon when I curl the first URL the server throws the error below:

```
curl http://localhost:4566/restapis/owe715e9gz/local/_user_request_/hello
{}
```

```
localstack_main   | 2021-04-09T18:46:12:WARNING:localstack.utils.server.http2_server: Error in proxy handler for request GET http://localhost:4566/restapis/owe715e9gz/local/_user_request_/hello: local variable 'payload' referenced before assignment Traceback (most recent call last):
localstack_main   |   File "/opt/code/localstack/localstack/utils/server/http2_server.py", line 117, in index
localstack_main   |     raise result
localstack_main   |   File "/opt/code/localstack/localstack/utils/bootstrap.py", line 654, in run
localstack_main   |     result = self.func(self.params)
localstack_main   |   File "/opt/code/localstack/localstack/utils/async_utils.py", line 28, in _run
localstack_main   |     return fn(*args, **kwargs)
localstack_main   |   File "/opt/code/localstack/localstack/services/generic_proxy.py", line 607, in handler
localstack_main   |     response = modify_and_forward(method=method, path=path_with_params, data_bytes=data, headers=headers,
localstack_main   |   File "/opt/code/localstack/localstack/services/generic_proxy.py", line 373, in modify_and_forward
localstack_main   |     listener_result = listener.forward_request(method=method,
localstack_main   |   File "/opt/code/localstack/localstack/services/edge.py", line 105, in forward_request
localstack_main   |     return do_forward_request(api, method, path, data, headers, port=port)
localstack_main   |   File "/opt/code/localstack/localstack/services/edge.py", line 116, in do_forward_request
localstack_main   |     result = do_forward_request_inmem(api, method, path, data, headers, port=port)
localstack_main   |   File "/opt/code/localstack/localstack/services/edge.py", line 136, in do_forward_request_inmem
localstack_main   |     response = modify_and_forward(method=method, path=path, data_bytes=data, headers=headers,
localstack_main   |   File "/opt/code/localstack/localstack/services/generic_proxy.py", line 373, in modify_and_forward
localstack_main   |     listener_result = listener.forward_request(method=method,
localstack_main   |   File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 48, in forward_request
localstack_main   |     return invoke_rest_api_from_request(method, path, data, headers)
localstack_main   |   File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 256, in invoke_rest_api_from_request
localstack_main   |     return invoke_rest_api(api_id, stage, method, relative_path_w_query_params,
localstack_main   |   File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 294, in invoke_rest_api
localstack_main   |     return invoke_rest_api_integration(api_id, stage, integration, method, path, invocation_path, data,
localstack_main   |   File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 394, in invoke_rest_api_integration
localstack_main   |     kwargs = {'name': payload['name']} if 'name' in payload else {}
localstack_main   | UnboundLocalError: local variable 'payload' referenced before assignment
```
Here is the function code:
```
def hello(event, context):
    return "Hello from lambda"
```

```
awslocal apigateway get-resources --rest-api-id owe715e9gz
{
    "items": [
        {
            "id": "xufs6xtjz0",
            "path": "/"
        },
        {
            "id": "av8whrnfz7",
            "parentId": "xufs6xtjz0",
            "pathPart": "hello",
            "path": "/hello",
            "resourceMethods": {
                "GET": {
                    "httpMethod": "GET",
                    "authorizationType": "NONE",
                    "apiKeyRequired": false,
                    "methodIntegration": {
                        "type": "AWS",
                        "httpMethod": "POST",
                        "uri": "arn:aws:apigateway:us-east-1:states:action/StartExecution",
                        "requestParameters": {},
                        "passthroughBehavior": "WHEN_NO_MATCH",
                        "cacheNamespace": "83d75f4a",
                        "cacheKeyParameters": [],
                        "integrationResponses": {
                            "200": {
                                "statusCode": 200,
                                "responseTemplates": {
                                    "application/json": null
                                }
                            }
                        }
                    }
                }
            }
        }
    ]
}
```

**Comment by whummer** [comment]
Thanks for reporting @Vmacedo-Daitan . The issue with `UnboundLocalError` should be resolved in the meantime - can you please pull the `latest` Docker image and give it another try? Please keep us posted how it goes. Thanks

**Comment by sjoukedv** [comment]
I'm also not able to call the endpoints, probably due to localstack proxy error:

```
2021-04-18T09:02:25:WARNING:localstack.utils.server.http2_server: Error in proxy handler for request GET http://localhost:4566/restapis/lllogmwuqb/local/_user_request_/gofunction: 'stateMachineArn' Traceback (most recent call last):

File "/opt/code/localstack/localstack/utils/server/http2_server.py", line 129, in index

raise result

File "/opt/code/localstack/localstack/utils/bootstrap.py", line 681, in run

result = self.func(self.params)

File "/opt/code/localstack/localstack/utils/async_utils.py", line 28, in _run

return fn(*args, **kwargs)

File "/opt/code/localstack/localstack/services/generic_proxy.py", line 352, in handler

response = modify_and_forward(method=method, path=path_with_params, data_bytes=data, headers=headers,

File "/opt/code/localstack/.venv/lib/python3.8/site-packages/localstack_ext/utils/aws/aws_utils.py", line 25, in modify_and_forward

File "/opt/code/localstack/localstack/services/generic_proxy.py", line 171, in modify_and_forward

listener_result = listener.forward_request(method=method,

File "/opt/code/localstack/localstack/services/edge.py", line 117, in forward_request

return do_forward_request(api, method, path, data, headers, port=port)

File "/opt/code/localstack/.venv/lib/python3.8/site-packages/localstack_ext/services/edge.py", line 167, in do_forward_request

return do_forward_request_orig(api,method,path,data,headers,*args,**kwargs)

File "/opt/code/localstack/localstack/services/edge.py", line 136, in do_forward_request

result = do_forward_request_inmem(api, method, path, data, headers, port=port)

File "/opt/code/localstack/localstack/services/edge.py", line 156, in do_forward_request_inmem

response = modify_and_forward(method=method, path=path, data_bytes=data, headers=headers,

File "/opt/code/localstack/localstack/services/generic_proxy.py", line 171, in modify_and_forward

listener_result = listener.forward_request(method=method,

File "/opt/code/localstack/.venv/lib/python3.8/site-packages/localstack_ext/services/apigateway/apigateway_extended.py", line 141, in forward_request

File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 256, in invoke_rest_api_from_request

return invoke_rest_api(api_id, stage, method, relative_path_w_query_params,

File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 294, in invoke_rest_api

return invoke_rest_api_integration(api_id, stage, integration, method, path, invocation_path, data,

File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 400, in invoke_rest_api_integration

result = client.start_execution(stateMachineArn=payload['stateMachineArn'],

KeyError: 'stateMachineArn'
```

However, using ` sls invoke stepf --stage local --name helloStateMachine` returns in the console

`[object Object]`

And LocalStack logs show the StateMachine execution was succesfull (ending with)

`2021-04-18 09:07:50.160: [200] DescribeExecution <= {"sdkResponseMetadata":null,"sdkHttpMetadata":null,"executionArn":"arn:aws:states:us-east-1:000000000000:execution:myHelloMachine:0b3236cc-46ba-4d1e-be6e-20a7e7255625","stateMachineArn":"arn:aws:states:us-east-1:000000000000:stateMachine:myHelloMachine","name":"0b3236cc-46ba-4d1e-be6e-20a7e7255625","status":"SUCCEEDED","startDate":1618736865110,"stopDate":1618736868955,"input":"{}","inputDetails":{"included":true},"output":"{\"body\":\"{\\n \\\"message\\\": \\\"Go Serverless v1.0! Your function executed successfully!\\\",\\n \\\"input\\\": {}\\n}\",\"statusCode\":200}\n","outputDetails":{"included":true},"traceHeader":null}`

Could it be that the http server that is handling the requests is misconfigured? 

LOCALSTACK_BUILD_DATE = 2021-04-15
LOCALSTACK_BUILD_GIT_HASH = c238c734

**Comment by Vmacedo-Daitan** [comment]
Thank you guys! 
I've updated the image and got the results below.

Deploy:
```
Serverless: Stack update finished...
Service Information
service: stepf-by-api
stage: local
region: us-east-1
stack: stepf-by-api-dev
resources: 12
api keys:
  None
endpoints:
  http://localhost:4566/restapis/hh4xur9ksq/local/_user_request_
functions:
  hello: stepf-by-api-local-hello
layers:
  None
Serverless StepFunctions OutPuts
endpoints:
  GET - https://hh4xur9ksq.execute-api.us-east-1.localhost/local/hello
```

CURL:
```
> curl http://localhost:4566/restapis/hh4xur9ksq/local/_user_request_/hello
{}
```

Log:
```
localstack_main   | 2021-04-23T18:23:50:WARNING:localstack.utils.server.http2_server: Error in proxy handler for request GET http://localhost:4566/restapis/hh4xur9ksq/local/_user_request_/hello: 'stateMachineArn' Traceback (most recent call last):
localstack_main   |   File "/opt/code/localstack/localstack/utils/server/http2_server.py", line 129, in index
localstack_main   |     raise result
localstack_main   |   File "/opt/code/localstack/localstack/utils/bootstrap.py", line 681, in run
localstack_main   |     result = self.func(self.params)
localstack_main   |   File "/opt/code/localstack/localstack/utils/async_utils.py", line 28, in _run
localstack_main   |     return fn(*args, **kwargs)
localstack_main   |   File "/opt/code/localstack/localstack/services/generic_proxy.py", line 352, in handler
localstack_main   |     response = modify_and_forward(method=method, path=path_with_params, data_bytes=data, headers=headers,
localstack_main   |   File "/opt/code/localstack/localstack/services/generic_proxy.py", line 171, in modify_and_forward
localstack_main   |     listener_result = listener.forward_request(method=method,
localstack_main   |   File "/opt/code/localstack/localstack/services/edge.py", line 117, in forward_request
localstack_main   |     return do_forward_request(api, method, path, data, headers, port=port)
localstack_main   |   File "/opt/code/localstack/localstack/services/edge.py", line 136, in do_forward_request
localstack_main   |     result = do_forward_request_inmem(api, method, path, data, headers, port=port)
localstack_main   |   File "/opt/code/localstack/localstack/services/edge.py", line 156, in do_forward_request_inmem
localstack_main   |     response = modify_and_forward(method=method, path=path, data_bytes=data, headers=headers,
localstack_main   |   File "/opt/code/localstack/localstack/services/generic_proxy.py", line 171, in modify_and_forward
localstack_main   |     listener_result = listener.forward_request(method=method,
localstack_main   |   File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 48, in forward_request
localstack_main   |     return invoke_rest_api_from_request(method, path, data, headers)
localstack_main   |   File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 256, in invoke_rest_api_from_request
localstack_main   |     return invoke_rest_api(api_id, stage, method, relative_path_w_query_params,
localstack_main   |   File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 294, in invoke_rest_api
localstack_main   |     return invoke_rest_api_integration(api_id, stage, integration, method, path, invocation_path, data,
localstack_main   |   File "/opt/code/localstack/localstack/services/apigateway/apigateway_listener.py", line 404, in invoke_rest_api_integration
localstack_main   |     result = client.start_execution(stateMachineArn=payload['stateMachineArn'],
localstack_main   | KeyError: 'stateMachineArn'
```

**Comment by pinzon** [comment]
Hi @Vmacedo-Daitan Thank you for being patient.  It took me some time to figure out how to fix the problem. 
With my latest patch, the stack deployment should work  and replicates the exact behaviour of AWS.
So I'll be closing the issue.

**Comment by whummer** [comment]
Thanks @Vmacedo-Daitan @pinzon . Should be fixed in #3957 - the `latest` Docker image is currently being built in CI and should become available in ~30 minutes from now. Thanks@
